pin/gc/gc.go

```diff
@@ -33,146 +27,232 @@ type Result struct {
-						continue loop
-					}
-					select {
-					case output <- Result{KeyRemoved: k}:
-					case <-ctx.Done():
-						break loop
-					}
-				}
-			case <-ctx.Done():
-				break loop
-			}
+		for _, c := range cids {
+			tri.InsertGray(c, r.Strict)
 		}
-		esweep.Append(logging.LoggableMap{
-			"whiteSetSize": fmt.Sprintf("%d", removed),
-		})
-		esweep.Done()
-		if errors {
-			output <- Result{Error: ErrCannotDeleteSomeBlocks}
-		}
-	}()
+	}
+	go g.gcAsync(ctx, emark, output, tri)
 
 	return output
 }
 
-func Descendants(ctx context.Context, getLinks dag.GetLinks, set *cid.Set, roots []*cid.Cid) error {
-	for _, c := range roots {
-		set.Add(c)
+func (g *gctype) gcAsync(ctx context.Context, emark *logging.EventInProgress,
+	output chan Result, tri *triset) {
 
-		// EnumerateChildren recursively walks the dag and adds the keys to the given set
-		err := dag.EnumerateChildren(ctx, getLinks, c, set.Visit)
-		if err != nil {
-			return err
+	defer close(output)
+
+	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
+		links, err := g.ls.GetLinks(ctx, cid)
+		if err != nil && err != dag.ErrNotFound {
+			return nil, &CannotFetchLinksError{cid, err}
 		}
+		return links, nil
 	}
 
-	return nil
-}
-
-// ColoredSet computes the set of nodes in the graph that are pinned by the
-// pins in the given pinner.
-func ColoredSet(ctx context.Context, pn pin.Pinner, ls dag.LinkService, bestEffortRoots []*cid.Cid, output chan<- Result) (*cid.Set, error) {
-	// KeySet currently implemented in memory, in the future, may be bloom filter or
-	// disk backed to conserve memory.
-	errors := false
-	gcs := cid.NewSet()
 	getLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
+		links, err := g.ls.GetLinks(ctx, cid)
 		if err != nil {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+			return nil, &CannotFetchLinksError{cid, err}
 		}
 		return links, nil
 	}
-	err := Descendants(ctx, getLinks, gcs, pn.RecursiveKeys())
+
+	var criticalError error
+	defer func() {
+		if criticalError != nil {
+			output <- Result{Error: criticalError}
+		}
+	}()
+
+	// Enumerate without the lock
+	for {
+		finished, err := tri.EnumerateStep(ctx, bestEffortGetLinks, getLinks)
+		if err != nil {
+			output <- Result{Error: err}
+			criticalError = ErrCannotFetchAllLinks
+		}
+		if finished {
+			break
+		}
+	}
+	if criticalError != nil {
+		return
+	}
+
+	// Add white objects
+	keychan, err := g.bs.AllKeysChan(ctx)
 	if err != nil {
-		errors = true
 		output <- Result{Error: err}
+		return
 	}
 
-	bestEffortGetLinks := func(ctx context.Context, cid *cid.Cid) ([]*node.Link, error) {
-		links, err := ls.GetLinks(ctx, cid)
-		if err != nil && err != dag.ErrNotFound {
-			errors = true
-			output <- Result{Error: &CannotFetchLinksError{cid, err}}
+loop:
+	for {
+		select {
+		case c, ok := <-keychan:
+			if !ok {
+				break loop
+			}
+			tri.InsertFresh(c)
```

Comment:
I think there's a bug here (well, not *here* but this is where I started writing the comment...). Specifically, what happens if I add a node (e.g., from the network) after the above `EnumerateStep` calls? I believe it will be garbage collected even if it shouldn't be.

The solution, as far as I can tell, is to leave unexplored gray nodes in the gray set and try to re-explore them after taking the GC lock.

Ideally, we'd have some nice way to subscribe to events from the blockstore so we could just process new blocks...

Reply:
`InsertFresh` in this inserts as white. I will change it to `InsertWhite` for time being. 

> what happens if I add a node (e.g., from the network) after the above EnumerateStep calls

If it is pinned it will be enumerated and marked black by enumeration when the lock is taken.

Reply:
You sure? I'm pretty sure those roots won't be re-enumerated because they're already black. I'm worried about the `a (pinned) -> b -> c (added late)` case. That is, `a` is a pin root and we only add `c` to the repo after the initial enumeration. I believe, in this case, we'll never mark `c` as black because `a` is already black.

Reply:
b can't be black as it references a white (or null, treated as white) node c.

If a is pinned strictly (by recursive pin) then marking b would fail with an error.
if a is pinned non-strictly (by Files API) then it is still better than it was before.

Reply:
> b can't be black as it references a white (or null, treated as white) node c.

No, if you have `b`, you'll definitely blacken it. `getLinks` will succeed because you have it (even if you don't have `c`). Yes, if you're pinning strictly you'll fail out.

>  if a is pinned non-strictly (by Files API) then it is still better than it was before.

It's still not correct. Actually, can't you just add the white nodes *first* (before you enumerate reachable nodes)? IIRC, that's how this algorithm usually works (for this reason).

Reply:
> No, if you have b, you'll definitely blacken it. getLinks will succeed because you have it (even if you don't have c). Yes, if you're pinning strictly you'll fail out.

Yes, then the `c` will be marked gray and processed in the next EnumerateStep.

> can't you just add the white nodes first (before you enumerate reachable nodes)?

I treat every node that doesn't exist in the tricolor set as white. They are being added here as it reduced time needed what we take the repo's GC lock.

Reply:
Not if it isn't strict. In that case, it will be marked gray, in the first `EnumerateStep` and then removed (and assumed to have no children).

> They are being added here as it reduced time needed what we take the repo's GC lock.

That doesn't mean you can't add them *before* the *first* enumerate step. Then, you'd follow the standard algorithm of moving nodes from the white set to the gray set to the black set until the gray set is empty.

Reply:
That is true, let's do that.